APIdefinition.md

```diff
@@ -0,0 +1,75 @@
+# ISIS Project public API
+
+The public API is considered to be all the interfaces that the ISIS
+project provides to the community from which the ISIS user base can
+build more complex tools. This API therefore includes code-level
+structures accessible via software calls to libraries, but also
+more user-facing structures like command-line programs and graphical
+user interface (GUI) programs. Any "interface" via which a user can
+reasonably interact with components of the ISIS project should be
+considered part of the "ISIS Project public API" for the purposes
+of [Semantic Versioning](https://semver.org).
+
+Examples are provided below, but are not meant to be exhaustive or
+complete. The above philosophy should guide our thinking in what
+the public API is.
+
+
+## General Guidance
+
+**Code-level structures**: Function, class, and object names,
+argument signatures, and return signatures for functions, classes,
+and objects.
+
+- C++ code exceptions would be those structures above marked with
+  a "private" access specifier via the C++ language.
+- Python code exceptions would be those structures above which begin 
+  with an underbar ("_", as per Python language tradition).
```

Comment:
Maybe we should add a note that the ABI is not a part of the public API. Users can expect to compile against new versions of the library without making changes, but cannot expect to link against new versions of the library without recompiling. For example, we may add new arguments, but they will have a default value.

Reply:
Can you provide suggested prose that I can paste in for what you'd like to see here?

Reply:
Maybe this is best handled by modifying this section and adding an example.

```
**Code-level structures**: Public function, member and class names,
argument signatures, and return signatures.

- In C++ code, public structures are structures that can be accessed by including the distributed header files for libraries.
- In Python code, public structure are structures that do not being with an underbar ("_"), as described in PEP 8
```

Then for the example:

```
### Adding new arguments

Adding new arguments to code level structures does not violate the existing public API as long as they provide a default value. For example changing a function doThing(int a, double b) to doThing(int a, double b, double c=1.0) does not violate the existing API for the doThing function as code that calls the doThing function will still compile without modification.

Adding new optional arguments or arguments with default values to command line programs does not violate the existing public API. For example, adding a new argument called "check" that has a default value of "false" to a command line program does not violate the program's public API as scripts and commands that were valid previously are still valid after the addition.
```

Reply:
Pushed a commit with this language added.